buildSrc/src/main/groovy/org/apache/fineract/gradle/service/GitService.groovy

```diff
@@ -0,0 +1,137 @@
+        GpgSigner.setDefault(new GpgSigner() {
+            @Override
+            void sign(@NonNull CommitBuilder commit, @Nullable String gpgSigningKey, @NonNull PersonIdent committer, CredentialsProvider credentialsProvider) throws CanceledException {
+                log.warn("------------------------ KEY: ${gpgSigningKey} IDENT: ${committer}")
+            }
+
+            @Override
+            boolean canLocateSigningKey(@Nullable String gpgSigningKey, @NonNull PersonIdent committer, CredentialsProvider credentialsProvider) throws CanceledException {
+                log.warn("------------------------ KEY: ${gpgSigningKey} IDENT: ${committer}")
+                return false
+            }
+        });
+
+        Repository repository = new FileRepositoryBuilder()
+                .readEnvironment()
+                .setGitDir(new File(config.dir))
+                .findGitDir()
+                .build();
+
+        this.git = new Git(repository)
+        this.dryRun = config.dryRun
+    }
+
+    void clone(String url, String branch, String directory) {
+        def ref = "refs/heads/" + branch
+
+        Git.cloneRepository()
+                .setURI(url)
+                .setDirectory(new File(directory))
+                .setBranchesToClone(Arrays.asList(ref))
+                .setBranch(ref)
+                .call();
+    }
+
+    void pull(String directory) {
+        Git git = Git.open(new File(directory));
+
+        git.pull().call()
+    }
+
+    void commit(String directory) {
+        Git git = Git.open(new File(directory));
+
+        def status = git.status().call()
+
+        if(!status.isClean()) {
+            git.add().addFilepattern(".").call()
+
+            git.commit().setSign(true).setMessage("chore: Publish current docs").call()
+        }
+
+        def result = git.push().call()
+    }
+
+    void config(String directory, List<FineractPluginExtension.FineractPluginConfigGitSection> sections) {
+        Git git = Git.open(new File(directory));
+
+        StoredConfig gitConfig = git.getRepository().getConfig();
+
+        sections.forEach(section -> {
+            gitConfig.setString(section.section, section.subsection, section.name, section.value);
+        })
+        // gitConfig.setString("branch", "master", "merge", "refs/heads/master");
+        gitConfig.save();
+    }
+
+    // TODO: explicitly provide the directory
+    void createTag(String name, String message) {
+        def ref = git.tag()
+                .setName(name)
+                .setMessage(message)
+                .setAnnotated(true)
+                .setForceUpdate(true)
+                .call()
+
+        try {
+            git.push().setPushTags().setForce(true).setDryRun(dryRun).call()
+
+            log.warn("Tag created: ${name} (${ref.name})")
+        } catch(Exception e) {
+            log.error(e.toString(), e)
```

Comment:
Don't we wanna rethrow an exception here?

Reply:
In any case the stacktrace will be written... if here or on the calling side doesn't matter that much. I might have another pass just concentrating on error management (there could be a need for an "undo" or "rollback" concept an error occurs in the middle of a release step.

Reply:
Yeah, I'm not only concerned about losing the stacktrace but stopping the execution in case there's an exception we're unprepared for.